File: src/jot/internal/splitter_ffi.mjs
export function make(patterns) {
  let pattern = "";
  let cursor = patterns;
  while (cursor.tail) {
    if (pattern !== "") pattern += "|";
    pattern += escapeRegExp(cursor.head);
    cursor = cursor.tail;
  }
  return new RegExp(pattern);
}

export function split(splitter, string) {
  const match = string.match(splitter);

  if (!match) return [string, "", ""]; // No delimiter found

  const index = match.index;
  const delimiter = match[0];

  return [
    string.slice(0, index),
    delimiter,
    string.slice(index + delimiter.length),
  ];
}

function escapeRegExp(string) {
  return string.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
}

File: src/jot/internal/splitter_ffi.erl
-module(splitter_ffi).
-export([new/1, split/2]).

new(Patterns) ->
    binary:compile_pattern(Patterns).

split(Splitter, String) ->
    case binary:match(String, Splitter) of
        nomatch -> {String, <<"">>, <<"">>};  % No delimiter found
        {Index, Length} ->
            {binary:part(String, 0, Index), 
             binary:part(String, Index, Length), 
             binary:part(String, Index + Length, byte_size(String) - Index - Length)}
    end.
